org.eclipse.m2e.core/src/org/eclipse/m2e/core/embedder/IMavenExecutionContext.java

```diff
@@ -106,4 +125,46 @@ public interface IMavenExecutionContext {
    */
   ProjectBuildingRequest newProjectBuildingRequest();
 
+  /**
+   * Join the currents thread {@link IMavenExecutionContext} because of this return an already executing context, this
+   * context can't be configured in any way, use {@link IMaven#createExecutionContext()} or
+   * {@link IMavenProjectFacade#createExecutionContext()} to create a specific context that could be configured.
+   * 
+   * @return the current {@link IMavenExecutionContext}
+   * @throws CoreException if no execution is attached to this thread
+   */
+  static IMavenExecutionContext join() throws CoreException {
+    IMavenExecutionContext context = MavenExecutionContext.getThreadContext();
+    if(context == null) {
+      throw new CoreException(Status.error("No active execution to join"));
+    }
+    return context;
+  }
+
+  /**
+   * Either join the currents thread {@link IMavenExecutionContext} or creates a fresh one from the supplied factory if
+   * this thread is currently not executing anything. Be aware that because of this might return an already executing
+   * context, this context should not be configured in any way, use {@link IMaven#createExecutionContext()} or
+   * {@link IMavenProjectFacade#createExecutionContext()} to create a specific context that could be configured.
+   * 
+   * @return a {@link IMavenExecutionContext} that could be used for execution
+   * @throws CoreException
+   */
+  static IMavenExecutionContext joinOrCreate(MavenExecutionContextFactory factory) throws CoreException {
+    if(canJoin()) {
+      return join();
+    }
+    return factory.createExecutionContext();
+  }
+
```

Comment:
I'm not sure all that should be static: at the moment, it is assumed that if there is a thread context, it's good to use it; but is it always true? ie if the current thread context doesn't use the same IMaven, is it the one that should be used?
I still think `static` is harmful here. Could those be non-static methods in the IMavenExecutionContext factory; or just a new non-static method in IMaven? I know you'd like to avoid it, but I think most ways around it are actually bringing more complexity and less flexiiblity. What is particularly wrong with `IMaven.context()` being an API that returns a suitable context (may reuse one from the thread or create one)?

Reply:
`MavenExecutionContext` is an internal class and as you can see exactly that method is used here :-)

Reply:
The method could be moved to `IMavenExecutionContext` if it's generally useful. And could be static as it really doesn't depend on any state beyond the static current thread.

Reply:
Well that's what this wrapper method (join()) do + some checks as interface can't contain (private) fields

Reply:
Yes, but then why an extra new method with more behavior if we can just expose an existing method? Seems to me that this method -as it takes a factory as input- does require extra interface and so on; while just exposing `getThreadContext()` wouldn't require so much.
Then, clients can do `Optional.of(IMavenExecutionContext.getThreadContext()).orElse(projectFacade::createExecutionContext)` which gives them more knowledge and responsibility over the context they want ot use. They may often realize that `projectFacade.createExecutionContext()` might be a better choice while this method would hide such possible improvements.

Reply:
> Yes, but then why an extra new method with more behavior if we can just expose an existing method? Seems to me that this method -as it takes a factory as input- does require extra interface and so on;

this is because you have had concerns using the `MavenPlugin().getMaven()` here is bad, so I added the opportunity to pass arbitrary "factory" method to construct a default.

> Then, clients can do

There is just one problem `createExecutionContext` throws an CoreException and thus you can't simply do as you suggested here (thats also why I needed a new interface in contrast to just using `Supplier<..>`.

We could change this, but this would be a (compile time) API break.